### PR TITLE
Fixing show typing and delay

### PR DESCRIPTION
**Pull request by RyanDawkins**
When the IActivitie's were being processed ServiceUrl and Conversation were not set. This was preventing the Typing from being sent.

I added the SendActivity method back because there is no flush method in order to force the Typing Activity to show in the chat client.

### CODE REVIEW

**Comment by RyanDawkins** [comment]
I don't like the type casting to Activity. I found another way of creating the Activity by setting the ServiceUrl and Conversation if that is preferred.

File: libraries/Microsoft.Bot.Builder/ContextExtensions/ContextExtensionMethods.cs [hunk 1]
```diff
@@ -2,20 +2,27 @@
...
         }
 
-        public static IBotContext Delay(this IBotContext context, int duration)
+        public static async Task<IBotContext> Delay(this IBotContext context, int duration)
         {
-            context.Responses.Add(new Activity { Type = "delay", Value = duration });
+            Activity activity = ((Activity)context.Request).CreateReply();
+            activity.Type = "delay";
+            activity.Value = duration;
+            await context.Bot.SendActivity(context, new List<IActivity>() { activity });
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/ContextExtensions/ContextExtensionMethods.cs]: I quite like that you've added these - I think they're very helpful. 

IF you can seperate them from the constant change in the Swagger layer, I think these are good to take. 

Now, to complicate things, I **think** we're going to change away from the queuing of responses, which are then run through a seperate adapter. It seems simpler all-the-way-around to simply have the Send methods just immediatly do a send. 

Doing that will make the framework work more how people expect, meaning we don't have to educate folks that "Send is not really send". That would be a big win. 

Until we make that change, however, (this week? next week?) these changes seem totally reasonable. 

Thanks~!!

Comment (RyanDawkins) [reply libraries/Microsoft.Bot.Builder/ContextExtensions/ContextExtensionMethods.cs]: By separating from swagger do you mean by removing the ActivityTypes.Delay?

Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder/ContextExtensions/ContextExtensionMethods.cs]: Yes. The way you added the ActivityTypes.Delay just won't work given our auto-generation of the .Schema package and related code. 

This code, except for that, looks very reasonable at first glance.

Comment (RyanDawkins) [reply libraries/Microsoft.Bot.Builder/ContextExtensions/ContextExtensionMethods.cs]: Okay cool. I have removed the Delay!

**Comment by cleemullins** [comment]
In case I didn't mention it, thank you. Your help here is really appreciated! 👍

**Comment by brandonh-msft** [comment]
I'd like to link this and #97 together.

**Comment by RyanDawkins** [comment]
@cleemullins Other than the ActivityTypes.Delay, is there anything else you are wanting me to change?

**Comment by cleemullins** [comment]
I’m not sure yet on this one. Let me review it again tonight.

________________________________

**Comment by brandonh-msft** [comment]
It's worth noting these changes removed the fluent-API capability of `ShowTyping()` and `Delay()`... we can no longer:
```csharp
context.Reply("msg 1").ShowTyping().Delay(2000).Reply("msg 2");
```
and must now
```csharp
context.Reply("msg 1");
await context.ShowTyping()
await context.Delay(2000);
context.Reply("msg 2");
```

**Comment by Stevenic** [comment]
@brandonh-msft that was by design as the feedback was they're confusing and further more they currently complicate the overall lifetime of the context object from a middleware perspective.  And Chris may not of applied all of the changes yet but the end goal for the base context object to only have:

```JavaScript
await context.sendActivities([
     { type: 'message', text: 'msg 1' },
     { type: 'typing' },
     { type: 'delay', value: 2000 },
     { type: 'message', text: 'msg 2' }
]);
```

Does that suck? yes it kind of does... But it's what you get with a unopinionated context object that's been stripped down to its bare basics.

The queue based approach can be added back in using middleware and a builder that looks something like this:

```JavaScript
Batch.add(context).reply(`msg 1`).typing().delay(2000).reply(`msg 2`);
```

I personally think it would be cleaner for the middleware to extend the context object like this:

```JavaScript
context.batch.reply(`msg 1`).typing().delay(2000).reply(`msg 2`);
```

But that would violate the immutable context object @billba has been pushing for.

**Comment by Stevenic** [comment]
So these were actually direct methods added to the `context` object, which again violates the immutability rules that @billba has been pushing for, so I'll let him comment on that as he believes pretty strongly that you shouldn't do stuff like this. For me its a case by case basis.

With that said... Yes these methods go direct and I think you'd want one approach or the other (either batch based or direct.)  There's a couple of advantages to the batch based approach over the direct approach (fluent style being one and potentially the ability to guarantee order on channels like kick) but I think having both models on the same context object will cause general confusion.  We've classically supported the batch based approach as it's the most flexible (or more accurately has the least issues) but I can understand why people hate having to manually call flush().

**Comment by Stevenic** [comment]
I should be clear that moving away from the batched based approach does introduce some issues in certain edge cases like Cortana skills but it's nothing they can't work around. We were just able to work around these issues for them before and now we can't.

**Comment by brandonh-msft** [comment]
> So these were actually direct methods added to the context object

I'm not sure if you mean "direct communication" or "direct-on-the-interface" … if the former, yes. if the latter, no - they're extension methods.

> I think having both models on the same context object will cause general confusion

100% agree.

**Comment by Stevenic** [comment]
@brandonh-msft I meant "direct communication"

**Pull request merged**